zerver/tests/test_subs.py

```diff
@@ -4127,6 +4131,33 @@ def test_json_get_subscribers(self) -> None:
             subscribers.append(subscriber)
         self.assertEqual(set(subscribers), set(expected_subscribers))
 
+    def test_json_get_subscribers_for_guest_user(self) -> None:
+        """
+        Guest users should have access to subscribers of web-public streams, even
+        if they aren't subscribed or have never subscribed to that stream.
+        """
+        guest_user = self.example_user("polonius")
+        _, _, never_subscribed = gather_subscriptions_helper(guest_user, True)
+        # A guest user can only see never subscribed streams that are web-public.
+        # For Polonius, the only web public stream that he is not subscribed at
+        # this point is Rome.
+        self.assertTrue(len(never_subscribed) == 1)
+
+        web_public_stream_id = never_subscribed[0]['stream_id']
+        expected_subscribers = never_subscribed[0]['subscribers']
+        result = self.client_get(f"/json/streams/{web_public_stream_id}/members")
+        self.assert_json_success(result)
+        result_dict = result.json()
+        self.assertIn('subscribers', result_dict)
+        self.assertIsInstance(result_dict['subscribers'], list)
+        subscribers: List[int] = []
+        for subscriber in result_dict['subscribers']:
+            self.assertIsInstance(subscriber, str)
+            subscriber = UserProfile.objects.get(email=subscriber)
+            subscribers.append(subscriber.id)
+
+        self.assertEqual(set(subscribers), set(expected_subscribers))
```

Comment:
We should make sure we assert the list isn't empty somewhere in this test; I might be happy to just do that (replacing the whole expected list generation dance).